src/analysis/convex/basic.lean

```diff
@@ -29,207 +28,388 @@ We also provide various equivalent versions of the definitions above, prove that
 are convex, and prove Jensen's inequality.
 
 Note: To define convexity for functions `f : E → β`, we need `β` to be an ordered vector space,
-defined using the instance `ordered_smul ℝ β`.
+defined using the instance `ordered_smul 𝕜 β`.
 
 ## Notations
 
-We use the following local notations:
-
-* `I = Icc (0:ℝ) 1`;
-* `[x, y] = segment x y`.
+We use the following notations:
+* `[x -[𝕜] y] = segment 𝕜 x y`
+* `I = Icc (0 : ℝ) 1`
 
-They are defined using `local notation`, so they are not available outside of this file.
+The second one is defined using `local notation`, so it is not available outside of this file.
 
 ## Implementation notes
 
 `convex_hull` is defined as a closure operator. This gives access to the `closure_operator` API
-while the impact on writing code is minimal as `convex_hull s` is automatically elaborated as
-`⇑convex_hull s`.
--/
+while the impact on writing code is minimal as `convex_hull 𝕜 s` is automatically elaborated as
+`⇑convex_hull 𝕜 s`.
 
-universes u' u v v' w x
+## TODO
 
-variables {E : Type u} {F : Type v} {ι : Type w} {ι' : Type x} {α : Type v'}
-  [add_comm_group E] [module ℝ E] [add_comm_group F] [module ℝ F]
-  [linear_ordered_field α]
-  {s : set E}
+Generalize all this file to affine spaces.
 
-open set linear_map
-open_locale classical big_operators pointwise
+Should we rename `segment` and `open_segment` to `convex.Icc` and `convex.Ioo`? Should we also
+define `clopen_segment`/`convex.Ico`/`convex.Ioc`?
+-/
 
-local notation `I` := (Icc 0 1 : set ℝ)
+universes u u'
+variables (𝕜 : Type*) {E F : Type*}
 
-section sets
+open linear_map set
+open_locale big_operators classical pointwise
 
 /-! ### Segment -/
 
+section ordered_semiring
+variables [add_comm_monoid E] [ordered_semiring 𝕜] [module 𝕜 E]
+
 /-- Segments in a vector space. -/
 def segment (x y : E) : set E :=
-{z : E | ∃ (a b : ℝ) (ha : 0 ≤ a) (hb : 0 ≤ b) (hab : a + b = 1), a • x + b • y = z}
+{z : E | ∃ (a b : 𝕜) (ha : 0 ≤ a) (hb : 0 ≤ b) (hab : a + b = 1), a • x + b • y = z}
 
-local notation `[`x `, ` y `]` := segment x y
+/-- Open segment in a vector space. Note that `open_segment 𝕜 x x = {x}` instead of being `∅` when
+the base semiring has some element between `0` and `1`. -/
+def open_segment (x y : E) : set E :=
+{z : E | ∃ (a b : 𝕜) (ha : 0 < a) (hb : 0 < b) (hab : a + b = 1), a • x + b • y = z}
+
+notation `[`x `-[` 𝕜`]` y `]` := segment 𝕜 x y
```

Comment:
```suggestion
notation `[`x ` -[` 𝕜`] ` y `]` := segment 𝕜 x y
```
I think should make the pretty printer output `[x -[𝕜] y]` instead of `[x-[𝕜]y]`?